deploy-tool/key/key_backup.go

```diff
@@ -0,0 +1,180 @@
+	// up secrets to a single GCP project. Fortunately the k8s secret names
+	// already include both the environment name and the namespace so no further
+	// disambiguation is needed.
+	return fmt.Sprintf("%s/secrets/%s", b.projectFragment(), name)
+}
+
+// containsSecret checks if a GCP secret corresponding to the provided
+// Kuberentes secret exists and returns:
+//   - secretDoesNotExist if the GCP secret does not exist at all
+//   - secretExists if the GCP secret exists but no enabled secret version was
+//	   found
+//	 - secretAndVersionExist if the GCP secret exists and an enabled secret
+//     version was found
+//
+// If something goes wrong or if a secret with an unexpected label is found,
+// returns secretDoesNotExist and an error.
+func (b *GCPSecretManagerBackup) containsSecret(k8sSecret *k8sv1.Secret) (containsSecretResult, error) {
+	// GCP Secret Manager has a *secret* resource to which labels may be
+	// attached and which contains *secret versions* which contain the actual
+	// secret values. First, we check if a secret with the expected name exists,
+	// and then verify that it has the expected label.
+	getSecretRequest := secretmanagerpb.GetSecretRequest{
+		Name: b.secretName(k8sSecret.ObjectMeta.Name),
+	}
+
+	gcpSecret, err := b.client.GetSecret(context.Background(), &getSecretRequest)
+	if err != nil {
+		if status.Code(err) == codes.NotFound {
+			return secretDoesNotExist, nil
+		}
+		return secretDoesNotExist, fmt.Errorf("failed to check if GCP secret %s exists: %w", b.secretName(k8sSecret.ObjectMeta.Name), err)
+	}
+
+	label, ok := gcpSecret.Labels[secretUIDLabel]
+	if !ok {
+		return secretDoesNotExist, fmt.Errorf(
+			"GCP secret %s does not have %s label - this should be resolved manually",
+			gcpSecret.Name, secretUIDLabel,
+		)
+	}
+	if label != string(k8sSecret.ObjectMeta.UID) {
+		return secretDoesNotExist, fmt.Errorf(
+			"label %s on GCP secret %s does not match Kubernetes secret UID %s - this should be resolved manually",
+			secretUIDLabel, gcpSecret.Name, k8sSecret.ObjectMeta.UID,
+		)
+	}
+
+	// Check if the secret contains at least one enabled version, which we
+	// assume to be our backed up secret.
+	listSecretVersionsRequest := secretmanagerpb.ListSecretVersionsRequest{
+		Parent: gcpSecret.Name,
+	}
+
+	versionsIterator := b.client.ListSecretVersions(context.Background(), &listSecretVersionsRequest)
+
+	for {
+		version, err := versionsIterator.Next()
+		if err == iterator.Done {
+			return secretExists, nil
+		} else if err != nil {
+			return secretDoesNotExist, fmt.Errorf("failed to list versions of secret %s", b.secretName(k8sSecret.ObjectMeta.Name))
+		}
+		if version.GetState() == secretmanagerpb.SecretVersion_ENABLED {
+			return secretAndVersionExist, nil
+		}
+	}
+}
+
+// BackupSecret backs up the provided Kubernetes secret into GCP Secret Manager,
+// if it is not already backed up.
+func (b *GCPSecretManagerBackup) BackupSecret(k8sSecret *k8sv1.Secret) error {
```

Comment:
This should take in a context argument I believe and pass it around to the calls that use a context argument.